File: pulse-react-native-otel/android/src/main/java/com/pulsereactnativeotel/ReactNativeScreenNameTracker.kt
package com.pulsereactnativeotel

import com.pulse.android.sdk.replay.SessionReplayController
import com.pulse.android.sdk.replay.SessionReplayRegistry
import java.util.concurrent.atomic.AtomicReference

/**
 * Allows React Native to override Android Activity/Fragment-based screen tracking.
 */
internal object ReactNativeScreenNameTracker {
    private val currentScreenName = AtomicReference<String?>()

    fun setCurrentScreenName(screenName: String?) {
        val previous = currentScreenName.getAndSet(screenName)
        if (previous != screenName) {
            val replay: SessionReplayController? = SessionReplayRegistry.getIntegration()
            replay?.notifyScreenChange()
        }
    }

    fun getCurrentScreenName(): String? = currentScreenName.get()
}

File: pulse-react-native-otel/ios/ReactNativeScreenNameTracker.swift
import Foundation
import PulseKit

/**
 * Allows React Native to override iOS ViewController-based screen tracking.
 */
@objc(ReactNativeScreenNameTracker)
public class ReactNativeScreenNameTracker: NSObject {
    private static let lock = NSLock()
    private static var _currentScreenName: String?
    private static var _previousScreenName: String?
    
    @objc public static func setCurrentScreenName(_ screenName: String?) {
        let changed: Bool
        lock.lock()
        if _currentScreenName != screenName {
            if let current = _currentScreenName {
                _previousScreenName = current
            }
            _currentScreenName = screenName
            changed = true
        } else {
            changed = false
        }
        lock.unlock()

        if changed {
            SessionReplayInstrumentation.getInstance()?.recorderInstance?.notifyScreenChange()
        }
    }
    
    static func getCurrentScreenName() -> String? {
        lock.lock()
        defer { lock.unlock() }
        return _currentScreenName
    }

    static func getPreviousScreenName() -> String? {
        lock.lock()
        defer { lock.unlock() }
        return _previousScreenName
    }
}

